File: leetcode/python/576-out-of-boundary-paths.py
# Author: RT
# Date: 2022-07-16T14:31:16.073Z
# URL: https://leetcode.com/problems/out-of-boundary-paths/


from functools import cache


class Solution:
    def findPaths(
        self, m: int, n: int, maxMove: int, startRow: int, startColumn: int
    ) -> int:
        M = 10**9 + 7
        dirs = ((1, 0), (-1, 0), (0, 1), (0, -1))

        @cache
        def dfs(i: int, j: int, count: int):
            if i < 0 or i >= m or j < 0 or j >= n:
                return 1
            if not count:
                return 0
            return sum(dfs(i + dx, j + dy, count - 1) for dx, dy in dirs) % M

        return dfs(startRow, startColumn, maxMove) % M

    def find_paths__dp_bottom_up(
        self, m: int, n: int, maxMove: int, startRow: int, startColumn: int
    ) -> int:
        M = 10**9 + 7
        dirs = ((1, 0), (-1, 0), (0, 1), (0, -1))
        # represent count of ways to reach cell (i, j) given x moves
        # iterate for move in 0..maxMove, if a cell's count > 0, accumulate to
        # adjacent cells, if adjacent cell(s) is out of bound, add to answer
        ans = 0
        dp = [[0] * n for _ in range(m)]
        # initial state: move = 0
        dp[startRow][startColumn] = 1
        for move in range(1, maxMove + 1):
            # all balls within a move happen at the same time
            temp = [[0] * n for _ in range(m)]
            for i in range(m):
                for j in range(n):
                    if not dp[i][j]:
                        continue
                    outbounds = (i == 0) + (i == m - 1) + (j == 0) + (j == n - 1)
                    ans = (ans + dp[i][j] * outbounds) % M
                    for dx, dy in dirs:
                        x, y = i + dx, j + dy
                        if 0 <= x < m and 0 <= y < n:
                            temp[x][y] += dp[i][j] % M
            dp = temp

        return ans

File: leetcode/rust/576-out-of-boundary-paths.rs
// Author: RT
// Date: 2022-07-16T15:58:22.327Z
// URL: https://leetcode.com/problems/out-of-boundary-paths/

impl Solution {
    pub fn find_paths(m: i32, n: i32, max_move: i32, start_row: i32, start_column: i32) -> i32 {
        let (m, n) = (m as usize, n as usize);
        const M: i32 = 1_000_000_007;
        let dirs: [(i32, i32); 4] = [(1, 0), (-1, 0), (0, 1), (0, -1)];

        let mut ans: i32 = 0;
        let mut dp = vec![vec![0; n]; m];
        dp[start_row as usize][start_column as usize] = 1;

        for _ in 0..max_move {
            let mut temp = vec![vec![0; n]; m];
            for i in 0..m {
                for j in 0..n {
                    if dp[i][j] == 0 {
                        continue;
                    }
                    let outbounds = (i == 0) as i32
                        + (i == m - 1) as i32
                        + (j == 0) as i32
                        + (j == n - 1) as i32;
                    // use loop instead of multiply to prevent overflow
                    for _ in 0..outbounds {
                        ans = ans % M + dp[i][j] % M;
                    }
                    for dir in dirs {
                        let x = (i as i32 + dir.0) as usize;
                        let y = (j as i32 + dir.1) as usize;
                        if (0..m).contains(&x) && (0..n).contains(&y) {
                            // to prevent overflow
                            temp[x][y] = temp[x][y] % M + dp[i][j] % M;
                        }
                    }
                }
            }
            dp = temp;
        }

        ans % M
    }
}
